File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/scheduling/SyncSchedulingWorker.java
package com.lasthopesoftware.bluewater.client.stored.scheduling;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;
import androidx.work.Constraints;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.ListenableWorker;
import androidx.work.Operation;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;
import androidx.work.WorkerParameters;

import com.annimon.stream.Stream;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.SettableFuture;
import com.lasthopesoftware.bluewater.client.connection.libraries.LibraryConnectionProvider;
import com.lasthopesoftware.bluewater.client.library.access.LibraryRepository;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.LibraryFileProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.parameters.FileListParameters;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.stringlist.LibraryFileStringListProvider;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemServiceFileCollector;
import com.lasthopesoftware.bluewater.client.stored.library.sync.CheckForSync;
import com.lasthopesoftware.bluewater.client.stored.library.sync.SyncChecker;
import com.lasthopesoftware.bluewater.client.stored.scheduling.constraints.SyncWorkerConstraints;
import com.lasthopesoftware.bluewater.client.stored.service.StoredSyncService;
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.response.VoidResponse;
import com.namehillsoftware.lazyj.AbstractSynchronousLazy;
import com.namehillsoftware.lazyj.CreateAndHold;

import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class SyncSchedulingWorker extends ListenableWorker {

	private static final String workName = MagicPropertyBuilder.buildMagicPropertyName(SyncSchedulingWorker.class, "");

	public static Operation scheduleSync(Context context) {
		final PeriodicWorkRequest.Builder periodicWorkRequest = new PeriodicWorkRequest.Builder(SyncSchedulingWorker.class, 3, TimeUnit.HOURS);
		periodicWorkRequest.setConstraints(constraints(context));
		return WorkManager.getInstance(context)
			.enqueueUniquePeriodicWork(workName, ExistingPeriodicWorkPolicy.REPLACE, periodicWorkRequest.build());
	}

	private static Constraints constraints(Context context) {
		final SharedPreferences manager = PreferenceManager.getDefaultSharedPreferences(context);
		return new SyncWorkerConstraints(manager).getCurrentConstraints();
	}

	public static Promise<Boolean> promiseIsScheduled(Context context) {
		return promiseWorkInfos(context)
			.then(workInfos -> Stream.of(workInfos).anyMatch(wi -> wi.getState() == WorkInfo.State.ENQUEUED));
	}

	private static Promise<List<WorkInfo>> promiseWorkInfos(Context context) {
		return new Promise<>(m -> {
			final ListenableFuture<List<WorkInfo>> workInfosByName = WorkManager.getInstance(context).getWorkInfosForUniqueWork(workName);
			m.cancellationRequested(() -> workInfosByName.cancel(false));
			workInfosByName.addListener(() -> {
				try {
					m.sendResolution(workInfosByName.get());
				} catch (ExecutionException | InterruptedException e) {
					m.sendRejection(e);
				}
			}, AsyncTask.THREAD_POOL_EXECUTOR);
		});
	}

	private final CreateAndHold<CheckForSync> lazySyncChecker = new AbstractSynchronousLazy<CheckForSync>() {
		@Override
		protected CheckForSync create() {
			return new SyncChecker(
				new LibraryRepository(context),
				new StoredItemServiceFileCollector(
					new StoredItemAccess(context),
					new LibraryFileProvider(new LibraryFileStringListProvider(LibraryConnectionProvider.Instance.get(context))),
					FileListParameters.getInstance()));
		}
	};

	private final Context context;

	public SyncSchedulingWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
		super(context, workerParams);
		this.context = context;
	}

	@NonNull
	@Override
	public ListenableFuture<Result> startWork() {
		final SettableFuture<Result> futureResult = SettableFuture.create();
		lazySyncChecker.getObject().promiseIsSyncNeeded()
			.then(new VoidResponse<>(isNeeded -> {
				if (isNeeded)
					StoredSyncService.doSync(context);
			}))
			.then(
				v -> futureResult.set(Result.success()),
				futureResult::setException);
		return futureResult;
	}

	@Override
	public void onStopped() {}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/scheduling/SyncSchedulingWorker.kt
package com.lasthopesoftware.bluewater.client.stored.scheduling

import android.content.Context
import android.os.AsyncTask
import androidx.work.*
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.SettableFuture
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.LibraryFileProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.parameters.FileListParameters
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.stringlist.LibraryFileStringListProvider
import com.lasthopesoftware.bluewater.client.browsing.library.access.LibraryRepository
import com.lasthopesoftware.bluewater.client.connection.session.ConnectionSessionManager
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemServiceFileCollector
import com.lasthopesoftware.bluewater.client.stored.library.sync.SyncChecker
import com.lasthopesoftware.bluewater.client.stored.scheduling.constraints.SyncWorkerConstraints
import com.lasthopesoftware.bluewater.client.stored.service.StoredSyncService
import com.lasthopesoftware.bluewater.settings.repository.access.ApplicationSettingsRepository
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder
import com.namehillsoftware.handoff.promises.Promise
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

class SyncSchedulingWorker(private val context: Context, workerParams: WorkerParameters) : ListenableWorker(context, workerParams) {
	private val lazySyncChecker = lazy {
			SyncChecker(
				LibraryRepository(context),
				StoredItemServiceFileCollector(
					StoredItemAccess(context),
					LibraryFileProvider(LibraryFileStringListProvider(ConnectionSessionManager.get(context))),
					FileListParameters.getInstance()))
	}

	override fun startWork(): ListenableFuture<Result> {
		val futureResult = SettableFuture.create<Result>()
		lazySyncChecker.value.promiseIsSyncNeeded()
			.then { isNeeded ->
				if (isNeeded) StoredSyncService.doSync(context)
				futureResult.set(Result.success())
			}
			.excuse { futureResult.setException(it) }
		return futureResult
	}

	override fun onStopped() {}

	companion object {
		private val workName = MagicPropertyBuilder.buildMagicPropertyName(SyncSchedulingWorker::class.java, "")
		@JvmStatic
		fun scheduleSync(context: Context): Promise<Operation> {
			return constraints(context).then { c ->
				val periodicWorkRequest = PeriodicWorkRequest.Builder(SyncSchedulingWorker::class.java, 3, TimeUnit.HOURS)
				periodicWorkRequest.setConstraints(c)
				WorkManager.getInstance(context)
					.enqueueUniquePeriodicWork(workName, ExistingPeriodicWorkPolicy.REPLACE, periodicWorkRequest.build())
			}
		}

		private fun constraints(context: Context): Promise<Constraints> {
			val applicationSettings = ApplicationSettingsRepository(context)
			return SyncWorkerConstraints(applicationSettings).currentConstraints
		}

		fun promiseIsScheduled(context: Context): Promise<Boolean> {
			return promiseWorkInfos(context)
				.then { workInfos -> workInfos.any { wi -> wi.state == WorkInfo.State.ENQUEUED } }
		}

		private fun promiseWorkInfos(context: Context): Promise<List<WorkInfo>> {
			return object : Promise<List<WorkInfo>>() {
				init {
					val workInfosByName = WorkManager.getInstance(context).getWorkInfosForUniqueWork(workName)
					respondToCancellation { workInfosByName.cancel(false) }
					workInfosByName.addListener({
						try {
							resolve(workInfosByName.get())
						} catch (e: ExecutionException) {
							reject(e)
						} catch (e: InterruptedException) {
							reject(e)
						}
					}, AsyncTask.THREAD_POOL_EXECUTOR)
				}
			}
		}
	}
}
